engine/table/src/main/java/io/deephaven/engine/table/impl/QueryTable.java

```diff
@@ -1547,32 +1546,11 @@ private Table whereInInternal(final Table rightTable, final boolean inclusion,
         return QueryPerformanceRecorder.withNugget(
                 "whereIn(rightTable, " + inclusion + ", " + matchString(columnsToMatch) + ")",
                 sizeForInstrumentation(), () -> {
-                    final Table distinctValues;
-                    final boolean setRefreshing = rightTable.isRefreshing();
-
-                    final String[] rightColumnNames = MatchPair.getRightColumns(columnsToMatch);
-                    final DataIndex rightIndex = DataIndexer.getDataIndex(rightTable, rightColumnNames);
-                    if (rightIndex != null) {
-                        // We have a distinct index table, let's use it.
-                        distinctValues = rightIndex.table();
-                    } else if (setRefreshing) {
-                        distinctValues = rightTable.selectDistinct(rightColumnNames);
-                    } else {
-                        final TableDefinition rightDef = rightTable.getDefinition();
-                        final boolean allPartitioning =
-                                Arrays.stream(rightColumnNames).allMatch(cn -> rightDef.getColumn(cn).isPartitioning());
-                        if (allPartitioning) {
-                            distinctValues = rightTable.selectDistinct(rightColumnNames);
-                        } else {
-                            distinctValues = rightTable.coalesce();
-                        }
-                    }
-
                     final DynamicWhereFilter dynamicWhereFilter =
-                            new DynamicWhereFilter((QueryTable) distinctValues, inclusion, columnsToMatch);
+                            new DynamicWhereFilter(rightTable, inclusion, columnsToMatch);
                     final Table where = whereInternal(dynamicWhereFilter);
-                    if (distinctValues.isRefreshing()) {
-                        where.addParentReference(distinctValues);
+                    if (rightTable.isRefreshing()) {
+                        where.addParentReference(rightTable);
                     }
```

Comment:
We need to be adding a parent reference to the `distinctValues` table, not  necessary `rightTable`. Since we're adding a parent ref to the filter, it should be enough to make sure that the filter imposes reachability, liveness, and dependent-satisfaction on the actual set table.